c++/src/kj/buffering.h

```diff
@@ -0,0 +1,203 @@
+  // a copy, or transfer ownership of the buffer and allocate a new one to replace it.
+
+  RetainableFdBufferRange(Badge<SplittableFdBufferRange>, FdRangeInputBuffer& parent,
+      ArrayPtr<byte> bytes, Array<AutoCloseFd> fds);
+  RetainableFdBufferRange(Badge<SplittableFdBufferRange>, Own<void> ownBytes,
+      ArrayPtr<byte> bytes, Array<AutoCloseFd> fds);
+
+private:
+  friend class SplittableFdBufferRange;
+
+  OneOf<FdRangeInputBuffer*, Own<void>> parentOrRetainedBytes;
+  ArrayPtr<byte> bytes;
+  Array<AutoCloseFd> fds;
+};
+
+class SplittableFdBufferRange final {
+  // A handle to access a (potentially partial) range of bytes in an `FdRangeInputBuffer`.
+  // Prefixes of the range can be split off and retained for long-term usage.
+
+public:
+  ArrayPtr<byte> getBytes() { return bytes; }
+
+  Own<RetainableFdBufferRange> split(size_t splitAt);
+  // Splits off the range before splitAt and returns it. The returned range is consumed such that
+  // it will not be returned as part of a range returned by further calls to `.get()`
+  //
+  // Splitting such that FDs from two ranges would be included is an error.
+
+  Promise<void> finishRange(
+      OneOf<AsyncInputStream*, AsyncIoStream*, AsyncCapabilityStream*> stream, size_t amount) {
+    KJ_SWITCH_ONEOF(stream) {
+      KJ_CASE_ONEOF(inputStream, AsyncInputStream*) {
+        return finishRange(*inputStream, amount);
+      }
+      KJ_CASE_ONEOF(ioStream, AsyncIoStream*) {
+        return finishRange(*ioStream, amount);
+      }
+      KJ_CASE_ONEOF(capStream, AsyncCapabilityStream*) {
+        return finishRange(*capStream, amount);
+      }
+    }
+    KJ_UNREACHABLE;
+  }
+  Promise<void> finishRange(
+      AsyncInputStream& stream, size_t amount);
+  Promise<void> finishRange(
+      AsyncCapabilityStream& stream, size_t amount);
+  // If the range contains an incomplete prefix of the data you need to read out of it, call
+  // `finishRange` to finish the range, providing the `amount` of bytes necessary to do so. If the
+  // range already contains fds, it is an error to read more.
+  //
+  // If the current range + `amount` does not fit in the buffer,
+  // a new buffer (owned by this range) will be allocated.
+  //
+  // As an optimization, if the current range + `amount` DOES fit in the buffer AND this range is not
+  // backed by an overflow buffer, the underlying `read` call will allow reading up until the end of
+  // the buffer (i.e more than `amount`). The underlying FD range will be considered complete after
+  // `amount` bytes are read, such that any FDs read after `amount` will be part of a new FD range.
+
+ SplittableFdBufferRange(Badge<FdRangeInputBuffer>,
+      OneOf<FdRangeInputBuffer*, Own<FdRangeInputBuffer>> parentOrOverflow, ArrayPtr<byte> bytes);
+
+private:
+  friend class FdRangeInputBuffer;
+
+  struct Empty{}; // Used for when an overflow buffer is split away;
+  OneOf<FdRangeInputBuffer*, Own<FdRangeInputBuffer>, Empty> parentOrOverflow;
```

Comment:
So I'm not absolutely certain I understand this, but I'm not in love with the concept of an overflow buffer. Could we do something like this:
- Truncate our current buffer to the end of the last finished message/read.
- Allocate a new buffer, copy everything after the last finished message/read and put it in a list of buffers along with the previous one.
- Continue using the tail buffer in the list.